common/src/main/java/io/pravega/common/io/serialization/VersionedSerializer.java

```diff
@@ -0,0 +1,654 @@
+         * @param target    The target object to apply the deserialization to.
+         * @throws IOException If an IO Exception occurred.
+         */
+        public void deserialize(RevisionDataInput dataInput, ReaderType target) throws IOException {
+            deserialize(dataInput.getBaseStream(), target);
+        }
+
+        /**
+         * Deserializes data from the given ArrayView into the given object.
+         *
+         * @param data   The ArrayView to deserialize from.
+         * @param target The target object to apply the deserialization to.
+         * @throws IOException If an IO Exception occurred.
+         */
+        public void deserialize(ArrayView data, ReaderType target) throws IOException {
+            deserialize(data.getReader(), target);
+        }
+
+        /**
+         * Deserializes data from the given InputStream into the given object.
+         *
+         * @param stream The InputStream to deserialize from.
+         * @param target The target object to apply the deserialization to.
+         * @throws IOException If an IO Exception occurred.
+         */
+        public void deserialize(InputStream stream, ReaderType target) throws IOException {
+            processHeader(stream);
+            deserializeContents(stream, target);
+        }
+
+        /**
+         * Deserializes data from the given InputStream into the given object. This does not attempt to read the serialization
+         * header.
+         *
+         * @param stream The InputStream to deserialize from.
+         * @param target The target object to apply the deserialization to.
+         * @throws IOException If an IO Exception occurred.
+         */
+        void deserializeContents(InputStream stream, ReaderType target) throws IOException {
+            DataInputStream dataInput = stream instanceof DataInputStream ? (DataInputStream) stream : new DataInputStream(stream);
+            byte version = dataInput.readByte();
+            val readVersion = this.versions[version];
+            ensureCondition(readVersion != null, "Unsupported version %d.", version);
+
+            byte revisionCount = dataInput.readByte();
+            ensureCondition(revisionCount >= 0, "Data corruption: negative revision count.");
+
+            int revisionIndex = 0;
+            for (int i = 0; i < revisionCount; i++) {
+                byte revision = dataInput.readByte();
+                val rd = readVersion.get(revisionIndex++);
+                try (RevisionDataInputStream revisionInput = RevisionDataInputStream.wrap(stream)) {
+                    if (rd != null) {
+                        // We've encountered an unknown revision; we cannot read anymore.
+                        ensureCondition(revision == rd.getRevision(),
+                                "Unexpected revision. Expected %d, found %d.", rd.getRevision(), revision);
+                        rd.getReader().accept(revisionInput, target);
```

Comment:
If I get this logic right, we are deserializing all revisions until we hit the last (`rd == null`) or the condition of the previous line fails. I think we need to do it because the revisions are stored in increasing order and consequently we need to traverse all revisions until we get to the last. Would it make sense to store in descending revision order instead?

Reply:
Revisions are incremental, which means that Revision N is built on Revisions 0..N-1. It makes more sense to store them in ascending order since that's how we'll be reading the data back; when we construct the object, we start with revision 0 (oldest), apply that, then go all the way up to N. You could achieve a similar effect by storing them backwards, though that really looks counter-intuitive to me. When we encounter a revision we do not recognize, we skip all future revisions until the end of the serialization.